framework/include/utils/Buffer.h

```diff
@@ -0,0 +1,275 @@
+
+  /**
+   * Create an empty buffer
+   */
+  Buffer();
+
+  /**
+   * Create a buffer with a specific capacity
+   */
+  Buffer(const std::size_t capacity);
+
+  virtual ~Buffer() {}
+
+  /**
+   * Resize the capacity
+   */
+  void setCapacity(const std::size_t capacity) { this->_data.resize(capacity); }
+  /**
+   * Get the capacity
+   */
+  std::size_t capacity() const { return this->_data.size(); }
+
+  /**
+   * Set the size
+   */
+  void setSize(const std::size_t size) { this->_end_pos = this->newEnd(this->_begin_pos + size); }
+  /**
+   * Get the size
+   */
+  std::size_t size() const { return this->_end_pos - this->_begin_pos; }
+
+  /**
+   * Whether or not the buffer is empty
+   */
+  bool empty() const { return !size(); }
+
+  /**
+   * Add a new entry on the end
+   */
+  void push_back(const T & value);
+
+  /**
+   * Moves the object into the buffer (calls std::move())
+   */
+  void move(T & value);
+
+  /**
+   * Add a new entries to the end
+   *
+   * Everything in [in_begin, in_end) is appended
+   */
+  void append(const_iterator in_begin, const_iterator in_end);
+  /**
+   * Add a new entries to the end
+   */
+  void append(const std::vector<T> & vals);
+
+  /**
+   * Remove all entries (does not change the capacity)
+   * Note: this does NOT at all free any entries
+   */
+  void clear();
+
+  /**
+   * Remove the first num elements
+   *
+   * Note that erased items are not guaranteed to be freed immediately
+   */
+  virtual void erase(const std::size_t num) = 0;
+  /*
+   * Similar to erase(), but if the chunk size is larger than the current size,
+   * no error (all elements are erased gracefully).
+   *
+   * Note that erased items are not guaranteed to be freed immediately
+   */
+  virtual void eraseChunk(const std::size_t chunk_size) = 0;
+
+  /**
+   * Iterator for the first entry in the buffer
+   */
+  iterator begin() { return this->_data.begin() + this->_begin_pos; }
+  /**
+   * Const iterator for the first entry in the buffer
+   */
+  const_iterator begin() const { return this->_data.begin() + this->_begin_pos; }
+
+  /**
+   * Iterator for the last entry in the buffer
+   */
+  iterator end() { return this->_data.begin() + this->_end_pos; }
+  /**
+   * Const iterator for the last entry in the buffer
+   */
+  const_iterator end() const { return this->_data.begin() + this->_end_pos; }
+
+  /**
+   * Iterator for the first entry with a given chunk size in the buffer
+   * If chunk_size is greater than the size of the buffer, the full range will be given.
+   */
+  virtual iterator beginChunk(const std::size_t chunk_size) = 0;
+  /**
+   * Const iterator for the first entry with a given chunk size in the buffer
+   * If chunk_size is greater than the size of the buffer, the full range will be given.
+   */
+  virtual const_iterator beginChunk(const std::size_t chunk_size) const = 0;
+
+  /**
+   * Iterator for the last entry with a given chunk size in the buffer
```

Comment:
```suggestion
   * Iterator for the last entry of a chunk of a given size in the buffer
```
I first understood this as "let's see the buffer as a bunch of chunks (of entries), return the position of the last chunk (so towards the end)"
If you have a clearer way of saying the end of the first chunk, down for that.